common/src/main/java/io/pravega/common/io/ByteBufferOutputStream.java

```diff
@@ -0,0 +1,183 @@
+ */
+package io.pravega.common.io;
+
+import io.pravega.common.io.serialization.RandomAccessOutputStream;
+import io.pravega.common.util.BufferView;
+import io.pravega.common.util.ByteArraySegment;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * An {@link OutputStream} that is backed by a {@link ByteBuffer} which can be resized as needed. Prefer using this
+ * instead of {@link java.io.ByteArrayOutputStream} since this makes direct use of {@link ByteBuffer} intrinsic methods
+ * for writing primitive types (see {@link DirectDataOutput} for example).
+ */
+@NotThreadSafe
+public class ByteBufferOutputStream extends OutputStream implements RandomAccessOutputStream {
+    //region Private
+
+    private static final int MAX_ARRAY_SIZE = Integer.MAX_VALUE - Long.BYTES;
+    private ByteBuffer buf;
+
+    //endregion
+
+    //region Constructor
+
+    /**
+     * Creates a new instance of the {@link ByteBufferOutputStream}.
+     */
+    public ByteBufferOutputStream() {
+        this(128);
+    }
+
+    /**
+     * Creates a new instance of the {@link ByteBufferOutputStream} with given initial size.
+     *
+     * @param initialSize The initial size to create with.
+     */
+    public ByteBufferOutputStream(int initialSize) {
+        this.buf = ByteBuffer.allocate(initialSize);
+    }
+
+    //endregion
+
+    //region Implementation
+
+    @Override
+    public void write(int b) {
+        ensureExtraCapacity(1);
+        this.buf.put((byte) b);
+    }
+
+    @Override
+    public void write(int byteValue, int streamPosition) {
+        this.buf.put(streamPosition, (byte) byteValue);
+    }
+
+    @Override
+    public void write(byte[] array) {
+        write(array, 0, array.length);
+    }
+
+    @Override
+    public void write(byte[] b, int offset, int length) {
+        ensureExtraCapacity(length);
+        this.buf.put(b, offset, length);
+    }
+
+    @Override
+    public void write(byte[] array, int arrayOffset, int length, int streamPosition) {
+        if (arrayOffset < 0 || length < 0 || (length > 0 && arrayOffset + length > array.length)) {
+            throw new ArrayIndexOutOfBoundsException("bufferOffset and length must refer to a range within buffer.");
+        }
+
+        final int extra = streamPosition + length - this.buf.position();
+        if (extra > 0) {
+            ensureExtraCapacity(extra);
+        }
+
+        final int originalPos = this.buf.position();
+        this.buf.position(streamPosition);
+        this.buf.put(array, arrayOffset, length);
+        this.buf.position(Math.max(originalPos, streamPosition + length));
+    }
+
+
+    @Override
+    public int size() {
+        return this.buf.position();
+    }
+
+    @Override
+    public void flush() {
+        // This method intentionally left blank.
+    }
+
+    @Override
+    public void close() {
+        // This method intentionally left blank.
```

Comment:
Have you thought about releasing the ByteBuffer with the cleaner ?

Reply:
Java NIO ByteBuffer does not have a release mechanism. Plus this is dealing exclusively with heap byte buffers which are recycled by the GC. Are you thinking of Netty ByteBuf?

Reply:
Probably I need a coffe :-) 
I thought you were using a direct memory buffer.
Sorry for the noise.

btw, are you using an heap buffer because the instances of this class won't be used directly for I/O ?

Reply:
This class is an equivalent of ByteArrayOutputStream but using ByteBuffer. I don't need it to be heap or direct, but heap is easier to deal with, is faster (to allocate) and I can always fetch a slice of the underlying array if needed (which is what the point of this class is anyway). 

I don't use it for IO directly, rather than as a way to conveniently serialize a set of objects.